### PR TITLE
[Dygraph]Add sharding stage2/3

### PR DESCRIPTION
**Pull request by Baibaifan**
### PR types

New features
### PR changes

APIs
### Description

[Dygraph]Add sharding stage2/3

```python
import paddle
from paddle.distributed.fleet.meta_optimizers.dygraph_optimizer.sharding_optimizer_stage2 import ShardingOptimizerStage2
from paddle.distributed.fleet.meta_parallel.sharding.sharding_stage2 import ShardingStage2
from paddle.distributed.fleet.meta_parallel.sharding.sharding_stage3 import ShardingStage3
fleet.init(is_collective=True)
group = paddle.distributed.new_group([0, 1])
# If use prue_fp16
scaler = paddle.amp.GradScaler(init_loss_scaling=scale_loss)
# wrap sharding scer
if sharding_stage in [2, 3]:
    scaler = ShardingScaler(scaler)
# wrap model
model = model_class(...)
# stage2
optimizer = ShardingOptimizer(params=model.parameters(), optim=optimizer, group=group)
model = ShardingStage2(model, oss_optimizer, group=group)
# stage3
model = ShardingStage3(model, optimizer=optimizer, group=group, offload=True)
# use optimizer as normal
img, label = data
label.stop_gradient = True
img.stop_gradient = True
out = model(img)
loss = paddle.nn.functional.cross_entropy(input=out, label=label)
loss.backward()
optimizer.step()
optimizer.clear_grad()
# stage3 if parameter need to convert to cpu, please add convert2cpu=True
model.get_all_parameters(convert2cpu=True)
```

### CODE REVIEW

**Review by ZHUI — COMMENTED** [review]
是否有 ShardingOptimizerStage 相关使用文档？

用户可能不清楚具体所指的功能

**Comment by Baibaifan** [comment]
> 是否有 ShardingOptimizerStage 相关使用文档？
> 
> 用户可能不清楚具体所指的功能

可以先参考：http://agroup.baidu.com/paddlepaddle/md/article/4646309

File: examples/language_model/gpt-3/dygraph/args.py [hunk 1]
```diff
@@ -224,6 +224,22 @@ def parse_args(MODEL_CLASSES):
         const=False,
         help="Using the recompute to save the memory.")
 
+    # add sharding stage2/3
+    parser.add_argument(
+        "--sharding_stage",
+        type=int,
+        default=1,
+        help="sharding stage1/2/3."
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/args.py]: 稍微补充一下，`stage1/2/3`的功能描述？

File: examples/language_model/gpt-3/dygraph/modeling.py [hunk 2]
```diff
@@ -793,6 +793,8 @@ def __init__(self, gpt):
         super(GPTForPretraining, self).__init__()
         self.gpt = gpt
         self.apply(self.init_weights)
+        # extra_parameters using for sharding stage3 to register extra_parameters
+        self.extra_parameters = [self.gpt.embeddings.word_embeddings.weight]
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/modeling.py:798]: 这个是写死的hard code?

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 3]
```diff
@@ -224,7 +230,7 @@ def do_train(args):
         if not any(nd in n for nd in ["bias", "norm"])
     ]
 
-    if args.sharding_degree > 1:
+    if args.sharding_stage == 1:
         optimizer = DygraphShardingOptimizer(
             hcg=fleet.get_hybrid_communicate_group(),
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:239]: sharding_stage1 等价于 DygraphShardingOptimizer ？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 4]
```diff
@@ -255,15 +261,30 @@ def do_train(args):
 
     if args.use_pure_fp16:
         scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
-        scaler = fleet.distributed_scaler(scaler)
         # level O2 means converting the network to FP16
-        model, optimizer = paddle.amp.decorate(
+        if args.sharding_stage in [2, 3]:
+            scaler = ShardingScaler(scaler)
+        else:
+            scaler = fleet.distributed_scaler(scaler)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: `fleet.distributed_scaler` == sharding with AMP?
`ShardingScaler` == harding with pure_fp16 ?

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 5]
```diff
@@ -255,15 +261,30 @@ def do_train(args):
...
-        scaler = fleet.distributed_scaler(scaler)
         # level O2 means converting the network to FP16
-        model, optimizer = paddle.amp.decorate(
+        if args.sharding_stage in [2, 3]:
+            scaler = ShardingScaler(scaler)
+        else:
+            scaler = fleet.distributed_scaler(scaler) 
+        model = paddle.amp.decorate(
             models=model,
-            optimizers=optimizer,
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:262]: optimizer和 amp 解耦了？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 6]
```diff
@@ -255,15 +261,30 @@ def do_train(args):
...
             level='O2',
             save_dtype='float32')
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: `level='O2'`
这里是默认 O2, 是否也添加一下`O1`支持？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 7]
```diff
@@ -255,15 +261,30 @@ def do_train(args):
...
-    if paddle.distributed.get_world_size() > 1:
+    # wrap sharding stage2/3 and add collective group
+    if args.sharding_stage in [2, 3]:
+        group = fleet.get_hybrid_communicate_group().get_check_parallel_group()
+        accumulate_grads = True if accumulate_steps > 1 else False
+    
+    if args.sharding_stage == 2:
+        optimizer = ShardingOptimizerStage2(params=model.parameters(), optim=optimizer,
+                                            group=group, offload=args.sharding_offload)
+        model = ShardingStage2(model, optimizer, group=group, accumulate_grads=accumulate_grads)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 为什么这里 `ShardingOptimizerStage2` 需要重新装饰一遍 optimizer ？Stage3 不用？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 8]
```diff
@@ -340,7 +361,10 @@ def do_train(args):
                         loss = loss + loss_mbs
 
                     if args.use_pure_fp16:
-                        scaler.minimize(optimizer, loss)
+                        if args.sharding_stage in [2, 3]:
+                            scaler.step(optimizer)
+                            scaler.update()
+                        else:scaler.minimize(optimizer, loss)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 我理解，这里应该是做了相同事情，可以接口统一吧

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 9]
```diff
@@ -406,14 +431,15 @@ def do_train(args):
...
-                    ) > 1 else model
+                    ) > 1 and args.sharding_stage not in [2, 3] else model
                     output_dir = os.path.join(args.output_dir,
                                               "step_%d" % global_step)
                     os.makedirs(output_dir, exist_ok=True)
 
                     logger.info("Save model to %s" % output_dir)
 
                     if args.pp_degree > 1:
+                        assert args.sharding_stage not in [2, 3], "Sharding stage2/3 not support pp_degree > 1."
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 请把这些判断，加入到 传入参数的阶段，不要在训练阶段去判断了。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 10]
```diff
@@ -424,6 +450,9 @@ def do_train(args):
                                 "model_state_mp_{:0>2d}_sharding_{:0>2d}_pp_{:0>2d}.pdopt".
                                 format(mp_rank, sharding_rank, pp_rank)))
                     else:
+                        if args.sharding_stage == 3:
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:450]: `args.sharding_stage == 3 or 2` ？

这里的save 逻辑，和 mp, pp，dp等的逻辑关系可能要明确一下，现在非常乱。

**Review by ZHUI — COMMENTED** [review]
分布式策略逐渐增多，整体代码的逻辑显得过于复杂了。

**Comment by ZHUI** [comment]
代码请使用 pre-commit 格式化

File: examples/language_model/gpt-3/dygraph/args.py (hunk 1 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/args.py]: 补充论文介绍

File: examples/language_model/gpt-3/dygraph/modeling.py (hunk 2 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/modeling.py:798]: 这部分现在先使用用户注册的方式，之后会开发半自动注册方式

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 3 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:239]: 是的

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 4 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: ShardingScaler是对sharding stage2/3 scaler 的封装

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 5 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:262]: 这个和optimizer没有关系，可以去除

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 6 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: sharding sharding stage2/3 对于AMP使用方法和之前保持一致

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 7 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 这个和实现细节有关

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 8 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已统一成相同接口

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 9 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已修改

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 10 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:450]: 已反馈，之后应该有规划修改

**Comment by Baibaifan** [comment]
> 代码请使用 pre-commit 格式化

已修改

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 11]
```diff
@@ -143,9 +149,13 @@ def do_train(args):
     global_rank = hcg.get_global_rank()
     mp_rank = hcg.get_model_parallel_rank()
     pp_rank = hcg.get_stage_id()
-    dp_rank = hcg.get_data_parallel_rank()
+    pp_rank = hcg.get_data_parallel_rank()
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 错误修改

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 12]
```diff
@@ -143,9 +149,13 @@ def do_train(args):
...
     sharding_rank = hcg.get_sharding_parallel_rank()
 
+    # sharding stage2/3 not support hybrid parallel
+    if args.sharding_stage in [2, 3]:
+        assert args.dp_degree == args.mp_degree == args.pp_degree == 1
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 可以添加报错信息，表明暂时不支持

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 13]
```diff
@@ -255,15 +265,38 @@ def do_train(args):
...
         scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
-        scaler = fleet.distributed_scaler(scaler)
         # level O2 means converting the network to FP16
-        model, optimizer = paddle.amp.decorate(
-            models=model,
-            optimizers=optimizer,
-            level='O2',
-            save_dtype='float32')
+        if args.sharding_stage in [2, 3]:
+            scaler = ShardingScaler(scaler)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:271]: 记个TODO吧， `combine ShardingScaler` and `fleet.distributed_scaler` in feature

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 10 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:450]: args.sharding_stage == 2 的情况是哪里处理的？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 14]
```diff
@@ -340,12 +373,14 @@ def do_train(args):
                         loss = loss + loss_mbs
 
                     if args.use_pure_fp16:
-                        scaler.minimize(optimizer, loss)
+                        scaler.step(optimizer)
+                        scaler.update()
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/amp/GradScaler_cn.html

之前与 @zhangbo9674 沟通过，下面是统一api接口用法，可以记录一下TODO。
`scaler.minimize(optimizer, scaled)  # update parameters`

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 10 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:450]: > 已反馈，之后应该有规划修改

也在代码里记一下 TODO 吧

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 15]
```diff
@@ -255,15 +265,38 @@ def do_train(args):
...
+
+    elif args.sharding_stage == 3:
+        model = ShardingStage3(
+            model,
+            optimizer,
+            group=group,
+            accumulate_grads=accumulate_grads,
+            sync_comm=args.use_recompute,
+            offload=args.sharding_offload)
+
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 这一大段，连同上面的 `args.sharding_stage == 1`, 封装到一个函数里面吧，这块sharding的逻辑单独弄一个函数处理。

```python
  if args.sharding_stage == 1:
        optimizer = DygraphShardingOptimizer(
            hcg=fleet.get_hybrid_communicate_group(),
```

File: examples/language_model/gpt-3/dygraph/args.py [hunk 16]
```diff
@@ -224,6 +224,20 @@ def parse_args(MODEL_CLASSES):
         const=False,
         help="Using the recompute to save the memory.")
 
+    # add sharding stage2/3
+    parser.add_argument(
+        "--sharding_stage",
+        type=int,
+        default=1,
+        help="sharding stage1/2/3. Reference:https://arxiv.org/abs/1910.02054.")
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/args.py]: stage1 for sharing optimizer params. stage2 for  optimizer、 gradients params,   stage3 for  optimizer、 gradients、model  params 

我记得不是很清楚，你这里写清楚一点吧

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 13 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:271]: 已添加

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 15 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 这段之后会合并到distributed_model中

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 10 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:450]: args.sharding_stage == 2 不需要特殊处理

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 11 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已回退

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 12 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已修改

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 14 above)
Comment (Baibaifan) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已回退

**Review by ForFishes — DISMISSED** [review]
LGTM

**Review by ZHUI — APPROVED** [review]
LGTM
